File: app/src/main/java/com/surveysyncengine/data/remote/api/FakeSurveyApiService.kt
package com.surveysyncengine.data.remote.api

import com.surveysyncengine.domain.error.SurveyHttpException
import com.surveysyncengine.domain.model.SurveyResponse
import kotlinx.coroutines.delay
import java.net.SocketTimeoutException
import java.net.UnknownHostException
import java.util.UUID
import java.util.concurrent.atomic.AtomicInteger

// ---------------------------------------------------------------------------
// FakeSurveyApiService
//
// Fully configurable mock that covers every scenario in the spec:
//
//  Scenario 2 (partial failure):
//    FakeSurveyApiService(serverErrorOnCallIndices = setOf(5))
//
//  Scenario 3 (network degradation at call 4):
//    FakeSurveyApiService(timeoutOnCallIndices = setOf(3, 4))
//
//  Scenario 5 (all error types):
//    Use FailurePlan with mixed entries.
//
// Call indices are 0-based and global across the service lifetime.
// Call reset() between test cases.
// ---------------------------------------------------------------------------
class FakeSurveyApiService(
    private val failurePlan: List<FailureInstruction> = emptyList(),
    private val simulatedNetworkDelayMs: Long = 10L,
) : SurveyApiService {

    private val callIndex = AtomicInteger(0)

    // Runtime-adjustable delay — lets tests slow the API after construction
    // without rebuilding the failure plan. Used in concurrent sync tests.
    private var runtimeDelayMs: Long? = null
    fun setDelay(ms: Long) { runtimeDelayMs = ms }

    override suspend fun uploadSurveyResponse(response: SurveyResponse): UploadResponseDto {
        delay(runtimeDelayMs ?: simulatedNetworkDelayMs)
        val index = callIndex.getAndIncrement()
        applyFailurePlan(index)
        return UploadResponseDto(
            serverId = "srv-${UUID.randomUUID()}",
            receivedAt = System.currentTimeMillis(),
        )
    }

    override suspend fun uploadAttachment(
        surveyResponseId: String,
        attachmentId: String,
        localFilePath: String,
        mimeType: String,
    ): AttachmentUploadDto {
        delay(simulatedNetworkDelayMs)
        return AttachmentUploadDto(
            attachmentId = attachmentId,
            serverUrl = "https://fake-storage.example.com/$attachmentId",
        )
    }

    private fun applyFailurePlan(index: Int) {
        val instruction = failurePlan.firstOrNull { it.callIndex == index } ?: return
        when (instruction.failureType) {
            FailureType.TIMEOUT -> throw SocketTimeoutException("Simulated timeout at call $index")
            FailureType.NO_NETWORK -> throw UnknownHostException("Simulated no network at call $index")
            FailureType.SERVER_ERROR_500 -> throw SurveyHttpException(500, "Internal Server Error")
            FailureType.SERVER_ERROR_503 -> throw SurveyHttpException(503, "Service Unavailable")
            FailureType.CLIENT_ERROR_400 -> throw SurveyHttpException(400, "Bad Request")
            FailureType.CLIENT_ERROR_422 -> throw SurveyHttpException(422, "Unprocessable Entity")
            FailureType.UNKNOWN -> throw RuntimeException("Simulated unknown error at call $index")
        }
    }

    fun reset() = callIndex.set(0)

    fun callCount(): Int = callIndex.get()
}

// ---------------------------------------------------------------------------
// DSL helpers for building failure plans clearly in tests
// ---------------------------------------------------------------------------
data class FailureInstruction(
    val callIndex: Int,
    val failureType: FailureType,
)

enum class FailureType {
    TIMEOUT,
    NO_NETWORK,
    SERVER_ERROR_500,
    SERVER_ERROR_503,
    CLIENT_ERROR_400,
    CLIENT_ERROR_422,
    UNKNOWN,
}

/** Convenience builder. */
fun failurePlan(vararg pairs: Pair<Int, FailureType>): List<FailureInstruction> =
    pairs.map { (index, type) -> FailureInstruction(index, type) }

File: ios/SurveySyncEngineIOS/Data/Remote/API/FakeSurveyApiService.swift
//
//  FakeSurveyApiService.swift
//  SurveySyncEngineIOS
//
import Foundation

public enum FailureType {
    case timeout, noNetwork
    case serverError500, serverError503
    case clientError400, clientError422
    case unknown
}

public struct FailureInstruction {
    public let callIndex: Int
    public let failureType: FailureType
    public init(_ callIndex: Int, _ failureType: FailureType) {
        self.callIndex = callIndex
        self.failureType = failureType
    }
}

public func buildFailurePlan(_ pairs: (Int, FailureType)...) -> [FailureInstruction] {
    pairs.map { FailureInstruction($0.0, $0.1) }
}

public struct UnknownTestError: Error, LocalizedError {
    public let message: String
    public var errorDescription: String? { message }
}

public class FakeSurveyApiService: SurveyApiService {
    private let failurePlan: [FailureInstruction]

    // FIX Bug 1: previously two fields existed — `simulatedNetworkDelayMs` (read
    // by Task.sleep) and `delayNanoseconds` (written by setDelay but never read).
    // setDelay updated the wrong field so the concurrent sync test's 200ms delay
    // was silently ignored, making the test non-deterministic.
    // Fix: one field `delayNanoseconds`. setDelay writes it; Task.sleep reads it.
    private var delayNanoseconds: UInt64

    private var _callIndex = 0
    private let lock = NSLock()

    public init(failurePlan: [FailureInstruction] = [], simulatedNetworkDelayMs: UInt64 = 10) {
        self.failurePlan = failurePlan
        self.delayNanoseconds = simulatedNetworkDelayMs * 1_000_000
    }

    public func uploadSurveyResponse(response: SurveyResponse) async throws -> UploadResponseDto {
        try await Task.sleep(nanoseconds: delayNanoseconds)   // reads the single field
        let index = getAndIncrementCallIndex()
        try applyFailurePlan(at: index)
        return UploadResponseDto(serverId: "srv-\(UUID().uuidString)",
                                 receivedAt: Int64(Date().timeIntervalSince1970 * 1000))
    }

    public func uploadAttachment(surveyResponseId: String, attachmentId: String,
                                  localFilePath: String, mimeType: String) async throws -> AttachmentUploadDto {
        try await Task.sleep(nanoseconds: delayNanoseconds)
        return AttachmentUploadDto(attachmentId: attachmentId,
                                   serverUrl: "https://fake-storage.example.com/\(attachmentId)")
    }

    /// Updates the per-call delay at runtime.
    /// FIX: now writes to the same field Task.sleep reads.
    public func setDelay(_ nanoseconds: UInt64) { self.delayNanoseconds = nanoseconds }
    public func reset()        { lock.withLock { _callIndex = 0 } }
    public func callCount() -> Int { lock.withLock { _callIndex } }

    private func applyFailurePlan(at index: Int) throws {
        guard let i = failurePlan.first(where: { $0.callIndex == index }) else { return }
        switch i.failureType {
        case .timeout:        throw URLError(.timedOut)
        case .noNetwork:      throw URLError(.notConnectedToInternet)
        case .serverError500: throw SurveyHttpException(httpCode: 500, message: "Internal Server Error")
        case .serverError503: throw SurveyHttpException(httpCode: 503, message: "Service Unavailable")
        case .clientError400: throw SurveyHttpException(httpCode: 400, message: "Bad Request")
        case .clientError422: throw SurveyHttpException(httpCode: 422, message: "Unprocessable Entity")
        case .unknown:        throw UnknownTestError(message: "Simulated unknown error at call \(index)")
        }
    }

    private func getAndIncrementCallIndex() -> Int {
        lock.lock(); defer { lock.unlock() }
        let c = _callIndex; _callIndex += 1; return c
    }
}

extension NSLocking {
    @discardableResult
    func withLock<T>(_ block: () -> T) -> T { lock(); defer { unlock() }; return block() }
}
